### PR TITLE
Support dictionary array for divide kernel

### LINKED ISSUE #1982: Support DictionaryArray in divide kernel
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Adding DictionaryArray support to `divide` kernel.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1982.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
The clippy errors seems not caused by this change. I will fix them separately if it is reproducible locally.

**Comment by viirya** [comment]
cc @sunchao This is more complicated.

File: arrow/src/compute/kernels/arithmetic.rs [hunk 1]
```diff
@@ -892,6 +984,18 @@ where
...
 
+/// Perform `left / right` operation on two arrays. If either left or right value is null
+/// then the result is also null. If any right hand value is zero then the result of this
+/// operation will be `Err(ArrowError::DivideByZero)`.
+pub fn divide_dyn(left: &dyn Array, right: &dyn Array) -> Result<ArrayRef> {
+    match left.data_type() {
+        DataType::Dictionary(_, _) => {
+            typed_dict_math_op!(left, right, |a, b| a / b, math_divide_checked_op_dict)
+        }
+        _ => typed_math_op!(left, right, |a, b| a / b, math_checked_divide_op),
```
Comment (viirya) [inline arrow/src/compute/kernels/arithmetic.rs:995]: I extracted the called op function as a parameter.

File: arrow/src/compute/kernels/arithmetic.rs [hunk 2]
```diff
@@ -126,44 +126,68 @@ where
...
+                if r.is_zero() {
+                    Err(ArrowError::DivideByZero)
                 } else {
-                    Ok(T::default_value())
+                    Ok(op(l, r))
                 }
-            },
-        );
+            } else {
+                Ok(T::default_value())
```
Comment (sunchao) [inline arrow/src/compute/kernels/arithmetic.rs:167]: curious why we are using default value here

File: arrow/src/compute/kernels/arithmetic.rs [hunk 3]
```diff
@@ -892,6 +984,18 @@ where
     return math_checked_divide_op(left, right, |a, b| a / b);
 }
 
+/// Perform `left / right` operation on two arrays. If either left or right value is null
+/// then the result is also null. If any right hand value is zero then the result of this
+/// operation will be `Err(ArrowError::DivideByZero)`.
+pub fn divide_dyn(left: &dyn Array, right: &dyn Array) -> Result<ArrayRef> {
+    match left.data_type() {
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/arithmetic.rs:991]: Yea, I will implement these supports later.

File: arrow/src/compute/kernels/arithmetic.rs [hunk 4]
```diff
@@ -671,6 +741,28 @@ where
...
+    right: &DictionaryArray<K>,
+    op: F,
+) -> Result<PrimitiveArray<T>>
+where
+    K: ArrowNumericType,
+    T: ArrowNumericType,
+    T::Native: One + Zero,
+    F: Fn(T::Native, T::Native) -> T::Native,
+{
+    math_dict_divide_checked_op!(left, right, op, PrimitiveArray<T>)
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/arithmetic.rs:763]: okay, inlined it.

File: arrow/src/compute/kernels/arithmetic.rs [hunk 5]
```diff
@@ -126,44 +126,68 @@ where
...
+///
+/// This function errors if:
+/// * the arrays have different lengths
+/// * there is an element where both left and right values are valid and the right value is `0`
+fn math_checked_divide_op_on_iters<T, F>(
+    left: impl Iterator<Item = Option<T::Native>>,
+    right: impl Iterator<Item = Option<T::Native>>,
+    op: F,
+    len: usize,
+    null_bit_buffer: Option<Buffer>,
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/arithmetic.rs:151]: We still need `null_bit_buffer`.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Thanks @sunchao

**Pull request merged**